processor/src/main/java/org/mapstruct/ap/internal/model/beanmapping/SourceReference.java

```diff
@@ -239,11 +238,24 @@ private Parameter getSourceParameterFromMethodOrTemplate(String parameterName )
             if ( isForwarded ) {
                 Parameter parameter = Parameter.getSourceParameter( templateMethod.getParameters(), parameterName );
                 if ( parameter != null ) {
-                    result = method.getSourceParameters()
-                                   .stream()
-                                   .filter( p -> p.getType().isAssignableTo( parameter.getType() ) )
-                                   .collect( Collectors.reducing( (a, b) -> null ) )
-                                   .orElse( null );
+
+                    // When forward inheriting we should find the matching source parameter by type
+                    // If there are multiple parameters of the same type
+                    // then we fallback to match the parameter name to the current method source parameters
+                    for ( Parameter sourceParameter : method.getSourceParameters() ) {
+                        if ( sourceParameter.getType().isAssignableTo( parameter.getType() ) ) {
+                            if ( result == null ) {
+                                result = sourceParameter;
+                            }
+                            else {
+                                // When we reach here it means that we found a second source parameter
+                                // that has the same type, then fallback to the matching source parameter
+                                // in the current method
+                                result = Parameter.getSourceParameter( method.getParameters(), parameterName );
```

Comment:
is there a situation where the parameter you fetch here is not assignable? It seems like you fetch a parameter again only based on its name.

Reply:
No that is not possible. This is needed because the source reference is from the template method, so the name of the parameter is the one in that method. The name in the mapping method might not be the same. Therefore we need to find it by type. However, if there are multiple parameters of the same type we fallback to finding it by name